### ISSUE TITLE
Add `return_counts` to `torch.unique`

### ISSUE THREAD

**Issue by ptrblck** [pytorch/pytorch#12598]
## 🚀 Feature
The current implementation of `torch.unique` does not have a `return_counts` argument, which returns the counts of all unique occurrences in the tensor.

## Motivation

The feature request is related to some users missing this option, e.g. [this thread](https://discuss.pytorch.org/t/how-to-efficiently-perform-averaging-across-predefined-groups-in-a-tensor/27091/2?u=ptrblck). A current workaround is to use something like this:
```
x = torch.tensor([1, 2, 2, 2, 3])
x_unique = x.unique(sorted=True)
x_unique_count = torch.stack([(x==x_u).sum() for x_u in x_unique])
```

Also, numpy supports this option, which should be a nice to have for new users. [np.unique](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.unique.html)
As this might be added as the last argument, no breaking changes would occur.

## Pitch

I could add this option to the current `torch.unique` implementation as I've recently added the `return_inverse` option and am still familiar with the code.

## Alternatives

An alternative would be to use the hacky code from above.

Let me know, if that sound good to you!
Also a bit unrelated, but we could also add the `return_index` option together to match the numpy function as close as possible, but that's probably another feature request.

Best,
ptrblck

**Comment by ezyang** [comment]
I support adding this feature. Tag me if you open a PR.

**Comment by ptrblck** [comment]
@ezyang I have a working CPU version running now.
Before continuing to integrate the code I would like to discuss what's the best way of doing so.

Currently I could easily implement the feature by adding some lines into [_unique_dim_cpu_impl](https://github.com/pytorch/pytorch/blob/a91f3338a00440648af5487f8103f8023f4dee0a/aten/src/ATen/native/Unique.cpp#L52) as we've implemented the unique function manually for the `dim` argument a while ago.
Currently there is another implementation without the `dim` argument using an `std::unordered_set` to get the unique values. I could also add the `return_counts` functionality to it, but I have the feeling that the dim and non-dim versions are diverging quite a bit which could make this code hard to maintain.

Do you think a refactoring is necessary, i.e. to combine the dim and non-dim version and add `return_counts` afterwards or are my concerns ungrounded? 

CC @colesbury

**Comment by colesbury** [comment]
It's definitely nicer to have non-diverging code paths, but it's not the most important thing. (Correctness, functionality, and performance are more important.)

If you think the refactor isn't too much work, go for it. But it doesn't sound like it's essential for this PR.

If you do have diverging code paths, just make sure that the test cases cover both code paths.

**Comment by ptrblck** [comment]
I'm sorry for the delay! 
As I currently don't have access to a GPU, I had to postpone this PR a bit.
I'll try to get a GPU machine as soon as possible.

**Comment by qizhuli** [comment]
`return_counts` would be an awesome feature to have! @ptrblck do you have the GPU version implemented?

**Comment by zasdfgbnm** [comment]
The fix is temporarily reverted now

**Comment by oswinoswin** [comment]
Documentation says it's working (https://pytorch.org/docs/stable/torch.html#torch.unique) but 
`torch.unique(torch.tensor([1, 1, 2]), return_counts=True)` fails (unique() got an unexpected keyword argument 'return_counts').

**Comment by ptrblck** [comment]
@oswinoswin Yes, @zasdfgbnm  implemented it, so I'm closing this issue.

**Comment by JosueCom** [comment]
```python
votes = tensor([[2, 2, 1],
                [1, 1, 2]])

uni, count = th.unique(votes, dim=1, return_counts=True)

### output
uni = tensor([[1, 2],
              [2, 1]])
count = tensor([1, 2])
```

When I run the code above, count returns a shape of (2) instead of (2, 2) like the output. How does unique works when dim is specified? I am confused with the result of the function.

**Comment by klauskades** [comment]
I also needed a moment to understand it. I actually expected a bincount for each column.

Assuming the tensor x:
```
x = torch.tensor([[0, 4, 3, 4],[0, 1, 3, 1],[0, 1, 4, 1],[0, 4, 3, 4]])
x
# output
tensor([[0, 4, 3, 4],
        [0, 1, 3, 1],
        [0, 1, 4, 1],
        [0, 4, 3, 4]])
```
Calculating torch.unique I expected something like this:
```
uni, count = torch.unique(x, dim=1, return_counts=True)
# output:
uni = ((tensor([0]), tensor([4, 1]), tensor([3, 4]), tensor([4, 1])) # cannot be a matrix to due dimensions, not sure how to represent this
count = tensor([[4, 0, 0, 0, 0],
                [0, 2, 0, 0, 2],
                [0, 0, 0, 3, 1],
                [0, 2, 0, 0, 2]])
```
However, what we get and what is correct is:
```
uni = tensor([[0, 3, 4],
              [0, 3, 1],
              [0, 4, 1],
              [0, 3, 4]])
count = tensor([1, 1, 2])
```

Actually, specifying `dim=1` seems not to mean that it will apply a bincount  for each column, but it will count the unique occurrences of the whole column within the matrix.

Unfortunately, torch.bincount works only for 1-dimensional tensors so far. In the future it would be nice to have a feature that gives me the bincount for each column of a matrix.
